### DISCUSSION TITLE
Use Fragment Identifier in `hx-get` etc. to Conveniently Specify `hx-select`

### DISCUSSION THREAD

**Discussion by TheDocTrier** [bigskysoftware/htmx#3066] [Ideas]
For example,

```html
<a hx-get="/contacts/detail.html#archive-btn">Click Me</a>
```

would be equivalent to

```html
<a hx-get="/contacts/detail.html" hx-select="#archive-btn">Click Me</a>
```

Would this conflict with any current behavior? The essay on [Template Fragments](https://htmx.org/essays/template-fragments/) references some Java code which appears to do exactly this:

```java
Contact c = getContact();
ChillTemplates.render("/contacts/detail.html#archive-ui", "contact", c);
```

Discussion #1055 mentions that Unpoly selects the `main` tag by default, which seems to be a more limited version of this idea. I think the general behavior I described above would make it even easier to use HTMX with existing endpoints and in static sites which may have difficulty specifying `HX-Reselect`.

**Comment by TheDocTrier** [comment]
Here is a contrived example which demonstrates the behavior, each button selects part of the page using the URL fragment identifier:

```html
<!DOCTYPE HTML>
<html>
  <head>
    <script src="https://unpkg.com/htmx.org@2.0.3" integrity="sha384-0895/pl2MU10Hqc6jd4RvrthNlDiE9U1tWmX7WRESftEDRosgxNsQG/Ze9YMRzHq" crossorigin="anonymous"></script>

    <script>
      htmx.defineExtension('get-select', {
        onEvent: function(name, evt) {
          if (name === 'htmx:beforeProcessNode') {
            var elt = evt.detail.elt;
            var a = elt.attributes["hx-get"] || elt.attributes["hx-post"];
            if (a) {
              var h = new URL(a.value, window.location).hash;
              if (h) {
                elt.setAttribute("hx-select", h);
              }
            }
          }
        }
      });
    </script>

    <style>
#red {
  border: red solid 4px;
  padding: 1ch;
}
#both {
  border: black solid 4px;
  padding: 1ch;
}
    </style>
  </head>

  <body hx-ext="get-select">
    <div id="red">
      <button hx-get="#red" hx-trigger="click" hx-target="closest div">
        Nest RED!
      </button>
    </div>
    <div id="both">
      <button hx-get="" hx-trigger="click" hx-target="closest div">
        Nest both!
      </button>
    </div>
  </body>
</html>
```

Edit: demonstrate both positive and negative case.

**Reply by Telroshan** [reply]
Hey, just to let you know that if you're interested in exploring this idea further, feel free to make a repository for your extension and submit a PR against our [community extensions list](https://github.com/bigskysoftware/htmx/blob/master/www/content/extensions/_index.md#community-extensions) to add it in!
We suggest following the structure of our own extensions (see the [extensions repo](https://github.com/bigskysoftware/htmx-extensions), but basically source + documentation and tests) but you're ofc free to do it as you please, at it'd be your repo after all 😉

**Reply by MichaelWest22** [reply]
One idea to improve this extension is to instead use
```javascript
      htmx.defineExtension('hash-select', {
        onEvent: function(name, evt) {
          if (name === 'htmx:beforeSwap') {
            var elt = evt.detail.elt;
            var a = elt.attributes["hx-get"] || elt.attributes["hx-post"];
            if (a) {
              var h = new URL(a.value, window.location).hash;
              if (h) {
                evt.detail.selectOverride=h;
              }
            }
          }
        }
      });
```
where we use the beforeSwap event and set the selectOverride directly instead so that we don't have to update the element itself.  The only issue I can see is that this will override any manual hx-select property applied which could cause confusion.  It may be possible to use htmx function to get the hx-select value via the internalAPI so we could handle this edge case better.
```javascript
      let intAPI
      htmx.defineExtension('hash-select', {
        init: function(apiRef) {
        intAPI = apiRef
      },
        onEvent: function(name, evt) {
          if (name === 'htmx:beforeSwap') {
            var elt = evt.detail.elt;
            var a = elt.attributes["hx-get"] || elt.attributes["hx-post"];
            if (a) {
              var h = new URL(a.value, window.location).hash;
              if (h && !intAPI.getClosestAttributeValue(elt, 'hx-select')) {
                evt.detail.selectOverride=h;
              }
            }
          }
        }
      });
```

**Reply by TheDocTrier** [reply]
I think using `htmx:beforeSwap` without leveraging the internal API is the clearer way. Seeing how HTMX already overwrites the `selectOverride` value multiple times for both provided context and `HX-Reselect`:

https://github.com/bigskysoftware/htmx/blob/69857dfdbb0278fb53b948b9e8053edbe6ad07a6/src/htmx.js#L4811-L4817

I think it would be simpler to state in the documentation for this extension that the given fragment identifier will apply to `selectOverride` and leave the resulting cascade up to internal HTMX behavior.

**Reply by TheDocTrier** [reply]
@Telroshan I am not quite sure I understand what interface non-legacy extensions should be designed to use. The (legacy) extensions in https://github.com/bigskysoftware/htmx-extensions leverage undocumented and unversioned internal APIs (unless I have missed something here https://htmx.org/extensions/building/ ). It does not appear that `selectOverride` is stable for extensions to use. Does that mean we should use something like `transformResponse` to add `HX-Reselect` to incoming requests based on `elt`?

**Reply by Telroshan** [reply]
@TheDocTrier I can indeed not find `selectOverride` in the documentation, but be assured that once we introduce a feature, we always keep it backwards compatible (unless we release a major version like htmx2, which isn't likely to happen ever again), so I'd say `selectOverride` is safe to use.
As for the internal API, it's true that we also lack documentation for it (PRs always welcome about that btw!), but we also keep it backwards compatible as we know that community extensions (including extensions we may not even be aware of) might rely on these features. So, when we introduce additional parameters or refactor exposed methods (be it by the public API or the internal one exposed to extensions), we make sure the old calling syntax would still work as it was before such updates.

Hope this helps!

**Reply by TheDocTrier** [reply]
@Telroshan thanks for the info! I published the extension at https://github.com/TheDocTrier/hash-select . Based on what you said, I ended up using the request config path to obtain the fragment identifier. This should be generic across every request attribute HTMX provides. @MichaelWest22 thanks for the improvements!

**Reply by Telroshan** [reply]
Nice, feel free to submit a PR against our [community extensions list](https://github.com/bigskysoftware/htmx/blob/master/www/content/extensions/_index.md#community-extensions) to add it in if you want btw!
